### PR TITLE
Pushing down OR filters

### PR DESCRIPTION
**Pull request by tiagokepe**
Hey all,

This PR pushes down OR filters following the rules below:

1) It pushes down only single column in the root OR:
`SELECT * FROM integers WHERE a=1 OR a=2 AND (a>1 OR b<5)`
image

2) It doesn't push down OR filters with multiple columns in the root OR, e.g.:
`SELECT * FROM integers WHERE a=1 OR b=2 AND (a>1 OR b<5)`
image

3) It early stops pushing down when finding an AND filter, e.g.:
`SELECT * FROM integers WHERE a=1 OR a=2 OR a=3 AND (a>4 OR b<5)`
image

4) It preserves AND filters only if there is a single column in the filters, e.g.:
`SELECT * FROM integers WHERE a=1 OR (a>3 AND a<5)`
image

P.S.: It's missing to apply simplification rules in the OR filters and remove the filter operator when possible, e.g., in the last example the filter could be removed.

### CODE REVIEW

**Comment by pdet** [comment]
It is also working with arrow scans? The code is already there :-)

**Comment by tiagokepe** [comment]
> It is also working with arrow scans? The code is already there :-)
I didn't run the arrow tests, but it seems to work [https://github.com/tiagokepe/duckdb/runs/4409150347?check_suite_focus=true#:~:text=All%20tests,30](url)

Are your arrow scans in the branch master expecting ORs pushed down?

File: test/optimizer/table_or_pushdown.test [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+4	4
+
+# test only OR filters
+query II
+EXPLAIN SELECT * FROM integers WHERE a=1 OR a>3 OR a<5
+----
+physical_plan	<REGEX>:.*SEQ_SCAN.*Filters: a=1 OR a>3 OR a<5|.*
+
+query II
+SELECT * FROM integers WHERE a=1 OR a>3 OR a<5 ORDER by a
```
Comment (Mytherin) [inline test/optimizer/table_or_pushdown.test:53]: Could we add more tests with more complex `AND/OR` chains? And tests with more complex functions/conditions (e.g. `a=1 IS NULL`, `concat(a=1, b=1)='11'`, etc) within OR/AND. Case expressions (`CASE WHEN x=7 THEN ...`) within conjunctions.

File: test/optimizer/table_or_pushdown.test [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+
+query II
+EXPLAIN SELECT * FROM strings WHERE s>'AAAAAAAAAAAAAAA' OR s<'AAAAAAAAAAAAAAA99999A'
+----
+physical_plan	<REGEX>:.*SEQ_SCAN.*Filters: s>AAAAAAAAAAAAAAA|.*OR s<AAAAAAAAAAAAAAA99999A.*
+
+
+query I
+SELECT * FROM strings WHERE s>'AAAAAAAAAAAAAAA' OR s<'AAAAAAAAAAAAAAA99999A'
```
Comment (Mytherin) [inline test/optimizer/table_or_pushdown.test:92]: Perhaps test `<>` as well?

File: test/optimizer/table_or_pushdown.test [hunk 3]
```diff
@@ -0,0 +1,96 @@
+# name: test/optimizer/table_or_pushdown.test
+# description: Test Table OR Filter Push Down
+# group: [optimizer]
+
+statement ok
```
Comment (Mytherin) [inline test/optimizer/table_or_pushdown.test:5]: Could we add some tests with other scanners that perform filter pushdown, e.g.:

* The Parquet scanner
* The Arrow scanner (as Pedro mentioned)

File: src/storage/table/column_segment.cpp [hunk 4]
```diff
@@ -279,15 +289,44 @@ static idx_t TemplatedNullSelection(SelectionVector &sel, idx_t approved_tuple_c
 	}
 }
 
-void ColumnSegment::FilterSelection(SelectionVector &sel, Vector &result, const TableFilter &filter,
-                                    idx_t &approved_tuple_count, ValidityMask &mask) {
+idx_t ColumnSegment::FilterSelection(SelectionVector &sel, Vector &result, const TableFilter &filter,
+                                     idx_t &approved_tuple_count, ValidityMask &mask) {
 	switch (filter.filter_type) {
+	case TableFilterType::CONJUNCTION_OR: {
+		// similar to the CONJUNCTION_AND, but we need to take cake of the SelectionVectors (OR all of them)
```
Comment (Mytherin) [inline src/storage/table/column_segment.cpp:296]: take care :) don't take my cake!

File: src/storage/table/column_segment.cpp [hunk 5]
```diff
@@ -210,6 +210,16 @@ static void FilterSelectionSwitch(T *vec, T *predicate, SelectionVector &sel, id
 		}
 		break;
 	}
+	case ExpressionType::COMPARE_NOTEQUAL: {
+		if (mask.AllValid()) {
```
Comment (Mytherin) [inline src/storage/table/column_segment.cpp:214]: This does not appear to be covered. Could you add some tests that trigger this?

File: src/storage/statistics/numeric_statistics.cpp [hunk 6]
```diff
@@ -119,6 +119,14 @@ FilterPropagateResult NumericStatistics::CheckZonemap(ExpressionType comparison_
 		} else {
 			return FilterPropagateResult::FILTER_ALWAYS_FALSE;
 		}
+	case ExpressionType::COMPARE_NOTEQUAL:
+		if (constant < min || constant > max) {
```
Comment (Mytherin) [inline src/storage/statistics/numeric_statistics.cpp:123]: Also uncovered

File: src/storage/statistics/string_statistics.cpp [hunk 7]
```diff
@@ -127,6 +127,11 @@ FilterPropagateResult StringStatistics::CheckZonemap(ExpressionType comparison_t
 		} else {
 			return FilterPropagateResult::FILTER_ALWAYS_FALSE;
 		}
+	case ExpressionType::COMPARE_NOTEQUAL:
+		if (min_comp < 0 || max_comp > 0) {
```
Comment (Mytherin) [inline src/storage/statistics/string_statistics.cpp:131]: Also uncovered

File: test/optimizer/table_or_pushdown.test [hunk 8]
```diff
@@ -0,0 +1,96 @@
+# name: test/optimizer/table_or_pushdown.test
+# description: Test Table OR Filter Push Down
+# group: [optimizer]
```
Comment (Mytherin) [inline test/optimizer/table_or_pushdown.test:3]: `PRAGMA enable_verification`

File: src/optimizer/filter_combiner.cpp [hunk 9]
```diff
@@ -987,4 +994,201 @@ ValueComparisonResult CompareValueInformation(ExpressionValueInformation &left,
...
+			auto child_conjunction = (BoundConjunctionExpression *)child.get();
+			conjunctions_to_visit.emplace_back(child_conjunction);
+			break;
+		}
+		case ExpressionClass::BOUND_COMPARISON: {
+			UpdateConjunctionFilter((BoundComparisonExpression *)child.get(), *child);
+			break;
+		}
+		default: {
+			SetEarlyStopPushdown(true);
```
Comment (Mytherin) [inline src/optimizer/filter_combiner.cpp:1038]: This stateful approach of using a function (`SetEarlyStopPushdown`) to set a flag somewhere that aborts the pushdown seems quite error-prone. Could we not just return from this function whether or not the pushdown has succeeded, and not push down the OR node at all if the pushdown has not succeeded (for whatever reason)?

File: src/optimizer/filter_combiner.cpp [hunk 10]
```diff
@@ -550,6 +553,8 @@ TableFilterSet FilterCombiner::GenerateTableScanFilters(vector<idx_t> &column_id
 		}
 	}
 
+	GenerateORFilters(table_filters, column_ids);
```
Comment (Mytherin) [inline src/optimizer/filter_combiner.cpp:556]: nit: `Or`, not `OR`, as it is not an abbreviation

File: src/optimizer/filter_combiner.cpp [hunk 11]
```diff
@@ -987,4 +994,201 @@ ValueComparisonResult CompareValueInformation(ExpressionValueInformation &left,
...
+		if (left_is_scalar || right_is_scalar) {
+			auto non_scalar_expr = left_is_scalar ? comparison_expr->right.get() : comparison_expr->left.get();
+			if (non_scalar_expr->GetExpressionType() == ExpressionType::BOUND_COLUMN_REF) {
+				UpdateFilterByColumn((BoundColumnRefExpression *)&expr, comparison_expr);
+				return;
+			}
+		}
+		SetEarlyStopPushdown(true);
+	}
+	ExpressionIterator::EnumerateChildren(expr,
```
Comment (Mytherin) [inline src/optimizer/filter_combiner.cpp:1080]: Should this iterate through the entire expression tree, even if the `SetEarlyStopPushdown` is triggered?

It also seems a bit odd to recursively iterate through the entire expression tree, e.g. what if I have something like `x=7 IS NULL`? Then `x` does not need to be equal to `7`, after all. Should this not only look at the elements that are actually contained in conjunctions?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Great that you got this to work. Some initial comments.

Could you also have a look at running some simple benchmarks that showcase performance improvements from the zonemap skipping this enables? e.g. a query like `SELECT SUM(l_orderkey) FROM lineitem WHERE l_orderkey=1 OR l_orderkey=80000` should be significantly faster with this change set.

File: src/optimizer/filter_combiner.cpp (hunk 9 above)
Comment (tiagokepe) [reply src/optimizer/filter_combiner.cpp:1038]: I think, at this point, it is possible to return and not push down the OR node.

**Comment by tiagokepe** [comment]
The last commits address all pointed out suggestions.
I added a new micro benchmark (benchmark/micro/pushdown/or_pushdown.benchmark) and the speed up with the or pushdown optimization is at least  1 order of magnitude as follows:

image

The result without the pushdown optimization (`SET disabled_optimizers TO 'filter_pushdown');
` is:

image

**Comment by Mytherin** [comment]
Thanks for the updates! Could you have a look at the various test failures and the codecov? There are still a number of lines that I would expect to be covered but that are not covered, e.g. `string_statistics.cpp:132`, `numeric_statistics.cpp:124/127`, `column_segment.cpp:218`, and many lines in the filter_combiner.

I'm also still not a big fan of the stateful manner in which the pushdown is stopped. Is the `stop_pushdown` variable necessary? Can't we do a pass where we check if the pushdown is possible with a separate function (e.g. `CanPushdownOr` or similar), and then bail-out if a pushdown is not possible?

In general having such conditions be dependent on an external state is quite error prone in my experience, as it is very easy to just ignore the variable and do something incorrect, or not correctly reset the variable, etc. It is better to avoid doing it in such a manner if possible.

**Comment by tiagokepe** [comment]
> Thanks for the updates! Very impressive results on the benchmark timings!

Yeah, I'm happy with the results, also thanks for your hints.

> 
> Could you have a look at the various test failures and the codecov? There are still a number of lines that I would expect to be covered but that are not covered, e.g. `string_statistics.cpp:132`, `numeric_statistics.cpp:124/127`, `column_segment.cpp:218`, and many lines in the filter_combiner.

For sure, I didn't have time to see the codecov, but I will check them.

> 
> I'm also still not a big fan of the stateful manner in which the pushdown is stopped. Is the `stop_pushdown` variable necessary? Can't we do a pass where we check if the pushdown is possible with a separate function (e.g. `CanPushdownOr` or similar), and then bail-out if a pushdown is not possible?
> 
> In general having such conditions be dependent on an external state is quite error prone in my experience, as it is very easy to just ignore the variable and do something incorrect, or not correctly reset the variable, etc. It is better to avoid doing it in such a manner if possible.

I think, with the current code, it's possible to remove the early_stop flag at all as the functions used during the BFSLookUpConjunctions() are returning false in case of finding any early stop pushdown condition, then the recursion process will stop. I need to check and apply minor refinements, but I guess we are almost there.

File: src/storage/statistics/string_statistics.cpp [hunk 12]
```diff
@@ -127,6 +127,11 @@ FilterPropagateResult StringStatistics::CheckZonemap(ExpressionType comparison_t
...
+			return FilterPropagateResult::FILTER_ALWAYS_TRUE;
```
Comment (Mytherin) [inline src/storage/statistics/string_statistics.cpp:132]: This still appears to be uncovered

**Review by Mytherin — COMMENTED** [review]
Thanks for the updates! This is looking great. One more minor request for tests, then this can be merged:

File: src/storage/statistics/string_statistics.cpp (hunk 12 above)
Comment (tiagokepe) [reply src/storage/statistics/string_statistics.cpp:132]: Done ;)

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**